src/qpx_mmc.cpp

```diff
@@ -0,0 +1,4244 @@
+  while (((drive->rd_buf[offs]) & 0x3F) != 0x2A)
+    offs++;
+  if (drive->rd_buf[offs + 6] & 0x02)
+    drive->parms.status |= STATUS_LOCK;
+  else
+    drive->parms.status &= (~STATUS_LOCK);
+  //#ifndef __PXCONTROL
+  //	if (!drive->silent) printf("--- Disc %slocked\n",(drive->parms.status &
+  // STATUS_LOCK) ? "" : "UN") #endif
+  return 0;
+}
+
+int set_lock(drive_info *drive) {
+  //	printf("--- %slocking disc...\n",(drive->parms.status & STATUS_LOCK) ?
+  //"" : "UN");
+  drive->cmd[0] = SPC_PREVENT_ALLOW_MEDIUM_REMOVAL;
+  drive->cmd[4] = (drive->parms.status & STATUS_LOCK) ? 1 : 0;
+  if ((drive->err = drive->cmd.transport(NONE, NULL, 0))) {
+    sperror("SET_LOCK", drive->err);
+    get_lock(drive);
+    return (drive->err);
+  }
+  get_lock(drive);
+  return 0;
+}
+
+int play_audio_msf(drive_info *drive, msf beg, msf end) {
+  drive->cmd[0] = MMC_PLAY_AUDIO_MSF;
+  drive->cmd[3] = beg.m;
+  drive->cmd[4] = beg.s;
+  drive->cmd[5] = beg.f;
+  drive->cmd[6] = end.m;
+  drive->cmd[7] = end.s;
+  drive->cmd[8] = end.f;
+  if ((drive->err = drive->cmd.transport(NONE, NULL, 0))) {
+    sperror("PLAY_AUDIO_MSF", drive->err);
+    return (drive->err);
+  }
+  return 0;
+}
+
+int play_audio(drive_info *drive, int32_t beg, short int len) {
+  drive->cmd[0] = MMC_PLAY_AUDIO;
+  drive->cmd[2] = (beg >> 24) & 0xFF;
+  drive->cmd[3] = (beg >> 16) & 0xFF;
+  drive->cmd[4] = (beg >> 8) & 0xFF;
+  drive->cmd[5] = beg & 0xFF;
+  drive->cmd[7] = (len >> 8) & 0xFF;
+  drive->cmd[8] = len & 0xFF;
+  if ((drive->err = drive->cmd.transport(NONE, NULL, 0))) {
+    sperror("PLAY_AUDIO", drive->err);
+    return (drive->err);
+  }
+  return 0;
+}
+
+int seek(drive_info *drive, int32_t lba, unsigned char flags) {
+  drive->cmd[0] = MMC_SEEK;
+  drive->cmd[2] = (lba >> 24) & 0xFF;
+  drive->cmd[3] = (lba >> 16) & 0xFF;
+  drive->cmd[4] = (lba >> 8) & 0xFF;
+  drive->cmd[5] = lba & 0xFF;
+
+  drive->cmd[9] = flags;
+  if ((drive->err = drive->cmd.transport(NONE, NULL, 0))) {
+    sperror("SEEK", drive->err);
+    return (drive->err);
+  }
+  return 0;
+}
+
+int read_cd(drive_info *drive, unsigned char *data, int32_t lba,
+            int sector_count, unsigned char flags, unsigned char FUA) {
+  //	int transfer_length = sector_count * 3072;
+
+  //	int sect_data = 2352;
+  int sect_data = 3072;
+
+  int transfer_length = sector_count * sect_data;
+  if (sector_count < 0)
+    return -1;
+
+  //	printf("lba: %d, cnt: %d\n", lba,sector_count);
+  drive->cmd[0] = MMC_READ_CD;
+  drive->cmd[1] = FUA ? 0x08 : 0x00;
+  drive->cmd[2] = (lba >> 24) & 0xFF;
+  drive->cmd[3] = (lba >> 16) & 0xFF;
+  drive->cmd[4] = (lba >> 8) & 0xFF;
+  drive->cmd[5] = lba & 0xFF;
+  drive->cmd[8] = sector_count;
+  drive->cmd[9] = flags;
+  if ((drive->err = drive->cmd.transport(READ, data, transfer_length))) {
+    sperror("READ_CD", drive->err);
+    return (drive->err);
+  }
+  return 0;
+}
+
+int read(drive_info *drive, unsigned char *data, int32_t lba, int sector_count,
```

Comment:
Codacy found an issue: [Check buffer boundaries if used in a loop including recursive loops (CWE-120, CWE-20).  ](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)